server/src/main/java/org/jetbrains/bsp/bazel/server/bep/BepServer.java

```diff
@@ -127,6 +126,39 @@ private void processCompletedEvent(BuildEventStreamProtos.BuildEvent event) {
     }
   }
 
+  private void processActionCompletedEvent(BuildEventStreamProtos.BuildEvent event) {
+    if (event.getId().hasActionCompleted()) {
+      consumeActionCompletedEvent(event);
+    }
+  }
+
+  private void consumeActionCompletedEvent(BuildEventStreamProtos.BuildEvent event) {
+    var label = event.getId().getActionCompleted().getLabel();
+    var actionEvent = event.getAction();
+    if (!actionEvent.getSuccess()) {
+      consumeUnsuccessfulActionCompletedEvent(actionEvent, label);
+    }
+  }
+
+  private void consumeUnsuccessfulActionCompletedEvent(
+      BuildEventStreamProtos.ActionExecuted actionEvent, String label) {
+    String stdErrText = "";
+    if (actionEvent.getStderr().getFileCase() == BuildEventStreamProtos.File.FileCase.URI) {
+      var path = Paths.get(URI.create(actionEvent.getStderr().getUri()));
+      try {
+        stdErrText = new String(Files.readAllBytes(path), StandardCharsets.UTF_8);
```

Comment:
could be replaced with `Files.readString()`

Reply:
ditto